consensus/core/src/leader_scoring.rs

```diff
@@ -75,9 +176,197 @@ impl ReputationScores {
+        );
+
+        assert!(
+            !blocks.is_empty(),
+            "Attempted to create UnscoredSubdag with no blocks"
+        );
+
+        Self {
+            context,
+            blocks,
+            commit_range,
+        }
+    }
+
+    pub(crate) fn get_leader_round_range(&self) -> (Round, Round) {
+        // Skip genesis round as we don't produce leaders for that round.
+        let first_round = self
+            .blocks
+            .keys()
+            .find(|block_ref| block_ref.round != 0)
+            .map(|block_ref| block_ref.round)
+            .expect("There should be a non-zero round in the set of blocks");
+        let last_round = self.blocks.keys().last().unwrap().round;
+        (first_round, last_round)
+    }
+
+    pub(crate) fn find_supported_block(
+        &self,
+        leader_slot: Slot,
+        from: &VerifiedBlock,
+    ) -> Option<BlockRef> {
+        if from.round() < leader_slot.round {
+            return None;
+        }
+        for ancestor in from.ancestors() {
+            if Slot::from(*ancestor) == leader_slot {
+                return Some(*ancestor);
+            }
+            // Weak links may point to blocks with lower round numbers than strong links.
+            if ancestor.round <= leader_slot.round {
+                continue;
+            }
+            if let Some(ancestor) = self.get_block(ancestor) {
+                if let Some(support) = self.find_supported_block(leader_slot, &ancestor) {
+                    return Some(support);
+                }
+            } else {
+                // TODO: Add unit test for this case once dagbuilder is ready.
+                tracing::info!(
+                    "Potential vote's ancestor block not found in unscored committed subdags: {:?}",
+                    ancestor
+                );
+                return None;
+            }
+        }
+        None
+    }
+
+    pub(crate) fn is_vote(
+        &self,
+        potential_vote: &VerifiedBlock,
+        leader_block: &VerifiedBlock,
+    ) -> bool {
+        let reference = leader_block.reference();
+        let leader_slot = Slot::from(reference);
+        self.find_supported_block(leader_slot, potential_vote) == Some(reference)
+    }
+
+    pub(crate) fn is_certificate(
+        &self,
+        potential_certificate: &VerifiedBlock,
+        leader_block: &VerifiedBlock,
+        all_votes: &mut HashMap<BlockRef, bool>,
+    ) -> bool {
+        let mut votes_stake_aggregator = StakeAggregator::<QuorumThreshold>::new();
+        for reference in potential_certificate.ancestors() {
+            let is_vote = if let Some(is_vote) = all_votes.get(reference) {
+                *is_vote
+            } else if let Some(potential_vote) = self.get_block(reference) {
+                let is_vote = self.is_vote(&potential_vote, leader_block);
+                all_votes.insert(*reference, is_vote);
+                is_vote
+            } else {
+                tracing::info!(
```

Comment:
This should be fairly common because of weak links and committed ancestors.
```suggestion
                tracing::trace!(
```